### DISCUSSION TITLE
Memory is complicated

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#20] [General]
# Memory is complicated

This is a small story about how memory operates in your .NET application. Well not only .NET but how memory does or does not get allocated.
We will see how a 1 Gigabyte big array is only a few megabytes big to some extend. Furthermore I will discuss working set and committed memory.

https://steven-giesel.com/blogPost/fee60a27-b1c4-491d-a2f1-3371db73ffac

**Comment by bpkinez** [comment]
Very interesting article! 

Thanks

**Comment by AloisKraus** [comment]
It is even more complicated. See https://aloiskraus.wordpress.com/2016/10/03/windows-10-memory-compression-and-more what else is making Task Manager unreliable if you are searching for memory leaks.

**Reply by linkdotnet** [reply]
Thanks for the link! Super informativ

**Comment by Patrick-3000** [comment]
Hello! You are talking about the fact, that reading from memory allocates physical memory. In your example, you are actually writing to memory. Does reading only (without writing) also allocate physical memory?

**Reply by linkdotnet** [reply]
Good point - reading means touching here. So in order to write to a certain location we have to read the address where we want to right. 
I'll use better words! Thanks!

**Comment by zen010101** [comment]
Here is a PowerShell script:

Get-Process -Name notepad |
    Select-Object -Property ID,
    @{Name = "WorkingSet(KB)"; Expression = {[math]::Round($_.WorkingSet64 / 1024, 2).ToString("#,##0")}},
    @{Name = "PrivateWS(KB)"; Expression = {[math]::Round($_.PrivateMemorySize64 / 1024, 2).ToString("#,##0")}},
    @{Name = "Commit(KB)"; Expression = {[math]::Round($_.PagedMemorySize64 / 1024, 2).ToString("#,##0")}} |
    Format-Table -AutoSize
